src/lakefs_spec/spec.py

```diff
@@ -422,14 +423,29 @@ def ls(self, path: str | os.PathLike[str], detail: bool = True, **kwargs: Any) -
         if info:
             # assumes that the returned info is name-sorted.
             pp = self._parent(info[0]["name"])
+            info_copy = info[:]
             if pp in self.dircache:
                 # ls info has files not in cache, so we update them in the cache entry.
-                cache_entry = self.dircache[pp]
-                # extend the entry by the new ls results
-                cache_entry.extend(info)
+                cache_entry = self.dircache[pp][:]
+
+                # Overwrite existing entries in the cache with its updated values
+                for idx, entry in enumerate(cache_entry):
+
+                    def _matches(e, name):
+                        return e["name"] == name
+
+                    info_entry = next(
+                        filter(partial(_matches, name=entry["name"]), info_copy),
+                        None,
+                    )
+                    if not info_entry:
+                        continue
+                    cache_entry[idx] = info_entry
+                    info_copy.remove(info_entry)
+                cache_entry.extend(info_copy)
```

Comment:
How about something like a JavaScript index?

```py
def safe_index(l, elem):
    try:
        return l.index(elem)
    except IndexError:
        return -1

for info_entry in info_copy:
    idx = safe_index(cache_entry, info_entry)
    if idx >= 0:
        cache_entry[idx] = info_entry
        info_copy.remove(info_entry)

cache_entry.extend(info_copy)

Reply:
Completely rewrote that part of the code in the meantime, please have a look :)